src/gui/model.rs

```diff
@@ -166,37 +182,95 @@ impl Greeter {
                 panic!("Failed to create session for username '{username}': {err}",)
             });
 
-        match response {
-            Response::Success => {
-                // No password is needed, so directly start the session.
-                info!("No password needed for current user");
-                self.start_session(sender, info);
-            }
-            Response::AuthMessage { auth_message, .. } => {
-                if auth_message.to_lowercase().contains("password") {
-                    // Show the password field, because a password is needed.
-                    self.updates.set_password_mode(true);
-                    self.updates.set_password(String::new());
-                } else {
-                    // greetd has requested something other than the password, so just display it
-                    // to the user and let them figure it out.
+        self.handle_greetd_response(sender, info, response);
+    }
+
+    /// This function handles a greetd response as follows:
+    /// - if the response indicates authentication success, start the session
+    /// - if the response is an authentication message:
+    ///     - for info and error messages (no input request), display/log the text and send an empty authentication response to greetd.
+    ///       This allows for immediate greetd updates when using authentication procedures that don't use text input.
```

Comment:
Could you clarify how the input (such as a fingerprint) is sent to greetd here? Does it automatically get it from the fingerprint drivers? If yes, would the loop be running infinitely until a fingerprint is given? If so, why not add a waiting time, so as not to unnecessarily keep looping too fast?

Also, what does an empty auth message signify?

Reply:
My information comes mostly from greetd-ipc(7) and inferred details from agreety. So, heads up, I didn't look through the greetd source myself (yet?), but this makes pretty good sense to me and so far it's been confirmed in my experiments.

So, this is how it works when using something like `fprint` in conjunction with greetd, from what I gather.

greetd tries an authentication step and therefore speaks indirectly to fprint. It gets a message like "Place your index finger on the fingerprint reader" and fprint then just waits for the driver, by itself. Now, greetd will give you this initial information via an authentication message of type info. If you now satisfy the request and make it read your fingerprint, authentication will be done but the greeter won't be finished because greetd doesn't send you a message without you sending one first. You'll have to send an authentication response to finish the process (that is why you have to press login twice currently).

The manpages say that an authentication response is always required:

> Indicates that an authentication message needs to be answered to con‐
tinue through the authentication flow. There are no limits on the num‐
ber and type of messages that may be required for authentication to 
succeed, and a greeter should not make any assumptions about the mes‐
ages. Must be answered with either post_auth_message_response or cancel_session.

So, when you respond with an empty authentication response immediately after receiving the info, greetd will wait for the authentication and you actually won't get a response until there is feedback from fprint. Thus, there actually is no CPU-intensive loop here because greetd makes you wait.

In other words, you could say that sending an empty authentication response is like reading a line of output from fprint/PAM, if that makes sense?

In short:
1. ReGreet -> greetd (create session)
2. greetd -> PAM -> fprint (authenticate)
3. greetd -> ReGreet (initial output from fprint)
4. Subsequent requests to greetd with empty authentication responses wait for fprint output, i.e. further info, failure or success
5. You use the scanner: fprint -> greetd (success/failure)
6. After the next (or currently waiting) request to greetd: greetd -> ReGreet (authentication success/error)

> Also, what does an empty message signify?
Hopefully my elaboration already cleared it up somehow, but I'll also quote greetd-ipc(7) again on empty responses (although it's not much of an explanation):

> If the message was informative (info, error), then a response does not need to be set in this message.

Reply:
Ah, got it. Thanks a lot for the detailed explanation!